File: app/src/main/java/com/site/vs/videostation/ui/category/CategoryActivity.java
package com.site.vs.videostation.ui.category;

import android.os.Bundle;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.TextView;

import com.site.vs.videostation.R;
import com.site.vs.videostation.base.MVPBaseActivity;
import com.site.vs.videostation.entity.CategoryDetailEntity;
import com.site.vs.videostation.widget.loadingtips.DefaultLoadingLayout;
import com.site.vs.videostation.widget.loadingtips.SmartLoadingLayout;

import java.util.HashMap;
import java.util.Map;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

/**
 * @author zhangbb
 * @date 2016/12/19
 */
public class CategoryActivity extends MVPBaseActivity<CategoryPresenter>
        implements CategoryDetailContract.View, CategoryFragment.FilterChangeListener {

    public static final String EXTRA_STRING_ID = "extra_id";
    public static final String EXTRA_STRING_CID = "extra_cid";

    @BindView(R.id.flayout_category)
    public FrameLayout frameLayout;
    @BindView(R.id.tv_title)
    public TextView tvTitle;
    private CategoryFragment categoryFragment;
    private DefaultLoadingLayout tipsLayout;
    private Map<String, String> paraMap;

    private String id;
    private String cid;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_category_homepage);
        ButterKnife.bind(this);
        tipsLayout = SmartLoadingLayout.createDefaultLayout(this, frameLayout);
        categoryFragment = new CategoryFragment();
        getSupportFragmentManager().beginTransaction().add(R.id.flayout_category, categoryFragment).commit();
        categoryFragment.setChangeListener(this);
        initData();
    }

    private void initData() {
        showLoading();
        id = getIntent().getStringExtra(EXTRA_STRING_ID);
        cid = getIntent().getStringExtra(EXTRA_STRING_CID);

        switch (Integer.parseInt(id)) {
            case 2:
                tvTitle.setText("电视剧");
                break;
            case 3:
                tvTitle.setText("动漫");
                break;
            case 4:
                tvTitle.setText("综艺");
                break;
        }

        paraMap = new HashMap<>();
        paraMap.put("cid", cid == null ? "" : cid);
        paraMap.put("year", "");
        paraMap.put("area", "");
        mPresenter.initCategoryBy(id, paraMap, 1, true);
    }

    @Override
    protected void createPresenter() {
        mPresenter = new CategoryPresenter();
        mPresenter.attachView(this);
    }

    @Override
    public void initCategorySuccess(CategoryDetailEntity entity) {
        tipsLayout.onDone();
        categoryFragment.setHeaderData(entity, cid);
        initCategoryContentSuccess(entity);
    }

    @Override
    public void initCategoryContentSuccess(CategoryDetailEntity entity) {
        if (entity != null && entity.list != null) {
            categoryFragment.updateList(entity);
        }
    }

    @Override
    public void loadMoreVideoSuccess(CategoryDetailEntity entity) {
        categoryFragment.appendList(entity);
    }

    @Override
    public void loadMoreVideoFail() {
        categoryFragment.loadMoreDataFail();
    }

    @Override
    public void initCategoryFail() {
        tipsLayout.onError();
        tipsLayout.setErrorButtonListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                mPresenter.initCategoryBy(id, paraMap, 1, true);
            }
        });
    }

    @Override
    public void showLoading() {
        tipsLayout.onLoading();
    }

    @Override
    public void hideLoading() {
        tipsLayout.onDone();
    }

    @Override
    public void filterChange(Map<String, String> map) {
        this.paraMap = map;
        mPresenter.initCategoryBy(id, map, 1, false);
    }

    @Override
    public void loadMore(int pageIndex) {
        if (paraMap != null)
            mPresenter.loadMoreVideo(id, paraMap, pageIndex);
    }

    @OnClick(R.id.iv_search)
    public void searchMovies() {
//        startActivity(new Intent(this, SearchActivity.class));
    }
}

File: app/src/main/java/com/site/vs/videostation/ui/category/CategoryActivity.kt
package com.site.vs.videostation.ui.category

import android.os.Bundle
import butterknife.ButterKnife
import com.site.vs.videostation.R
import com.site.vs.videostation.base.MVPBaseActivity
import com.site.vs.videostation.entity.CategoryDetailEntity
import com.site.vs.videostation.widget.loadingtips.DefaultLoadingLayout
import com.site.vs.videostation.widget.loadingtips.SmartLoadingLayout
import kotlinx.android.synthetic.main.activity_category_homepage.*
import java.util.*

/**
 * @author zhangbb
 * @date 2016/12/19
 */
class CategoryActivity : MVPBaseActivity<CategoryPresenter>(), CategoryDetailContract.View, CategoryFragment.FilterChangeListener {

    private var categoryFragment: CategoryFragment? = null
    private var tipsLayout: DefaultLoadingLayout? = null
    private var paraMap: MutableMap<String, String>? = null


    private var tid: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_category_homepage)
        ButterKnife.bind(this)
        tipsLayout = SmartLoadingLayout.createDefaultLayout(this, flayout_category)
        categoryFragment = CategoryFragment()
        supportFragmentManager.beginTransaction().add(R.id.flayout_category, categoryFragment!!).commit()
        categoryFragment!!.setChangeListener(this)
        initData()
    }

    private fun initData() {
        showLoading()
        tid = intent.getStringExtra(EXTRA_STRING_ID)

        when (Integer.parseInt(tid!!)) {
            29 -> tv_title!!.text = "电视剧"
            31 -> tv_title!!.text = "动漫"
            30 -> tv_title!!.text = "综艺"
        }

        paraMap = HashMap()
        paraMap!!["tid"] = "$tid"
        paraMap!!["year"] = ""
        paraMap!!["area"] = ""
        mPresenter.initCategoryBy(tid, paraMap, 1, true)



    }

    override fun createPresenter() {
        mPresenter = CategoryPresenter()
        mPresenter.attachView(this)
    }

    override fun initCategorySuccess(entity: CategoryDetailEntity) {
        tipsLayout!!.onDone()
        categoryFragment!!.setHeaderData(entity, tid)
        initCategoryContentSuccess(entity)
    }

    override fun initCategoryContentSuccess(entity: CategoryDetailEntity?) {
        if (entity != null && entity.category_list != null) {
            categoryFragment!!.updateList(entity)
        }
    }

    override fun loadMoreVideoSuccess(entity: CategoryDetailEntity) {
        categoryFragment!!.appendList(entity)
    }

    override fun loadMoreVideoFail() {
        categoryFragment!!.loadMoreDataFail()
    }

    override fun initCategoryFail() {
        tipsLayout!!.onError()
        tipsLayout!!.setErrorButtonListener { mPresenter.initCategoryBy(tid, paraMap, 1, true) }
    }

    override fun showLoading() {
        tipsLayout!!.onLoading()
    }

    override fun hideLoading() {
        tipsLayout!!.onDone()
    }

    override fun filterChange(map: MutableMap<String, String>) {
        this.paraMap = map
        mPresenter.initCategoryBy(tid, map, 1, false)
    }

    override fun loadMore(pageIndex: Int) {
        if (paraMap != null)
            mPresenter.loadMoreVideo(tid, paraMap, pageIndex)
    }

    fun searchMovies() {
        //        startActivity(new Intent(this, SearchActivity.class));
    }

    companion object {

        val EXTRA_STRING_ID = "extra_id"
        val EXTRA_STRING_CID = "extra_cid"
    }
}
